### PR TITLE
buildah: Allow secrets to be passed into builds safely

### PR DESCRIPTION
**Pull request by jimmyjones2**
# Changes

buildah allows secrets to be passed in a safe way that will not end up in the final image. Add this support to buildah Task:

```
FROM registry.access.redhat.com/ubi8/ubi
RUN --mount=type=secret,id=mysecret cat /run/secrets/mysecret
```

```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: use-secret-
spec:
  pipelineSpec:
    workspaces:
      - name: shared-workspace
      - name: secrets-ws
        optional: true
    tasks:
      - name: buildah
        taskRef:
          name: buildah
        workspaces:
        - name: source
          workspace: shared-workspace
        - name: secrets
          workspace: secrets-ws
        params:
        - name: IMAGE
          value: docker.io/library/foo
        - name: BUILD_EXTRA_ARGS
          value: --secret=id=mysecret,src=$(workspaces.secrets.path)/config.json
  workspaces:
    - name: shared-workspace
      configMap:
        name: dockerfile
    - name: secrets-ws
      secret:
        secretName: mysecret
```

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Follows the [authoring recommendations][authoring]
- [ ] Includes [docs][docs] (if user facing)
- [ ] Includes [tests][tests] (for new tasks or changed functionality)
  - See the [end-to-end testing documentation][e2e] for guidance and CI details.
- [x] Meets the [Tekton contributor standards][contributor] (including functionality, content, code)
- [x] Commit messages follow [commit message best practices][commit]
- [x] Has a kind label. You can add one by adding a comment on this PR that
  contains `/kind <type>`. Valid types are bug, cleanup, design, documentation,
  feature, flake, misc, question, tep
- [x] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [x] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [x] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [x] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [x] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [x] mandatory `spec.description` follows the convention

          ```

          spec:
            description: >-
              one line summary of the resource

              Paragraph(s) to describe the resource.
          ```

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md) for more details._

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]: https://github.com/tektoncd/catalog/issues/413
[authoring]: https://github.com/tektoncd/catalog/blob/main/recommendations.md
[docs]: https://github.com/tektoncd/community/blob/master/standards.md#docs
[tests]: https://github.com/tektoncd/community/blob/master/standards.md#tests
[e2e]: https://github.com/tektoncd/catalog/blob/main/CONTRIBUTING.md#end-to-end-testing
[contributor]: https://github.com/tektoncd/community/blob/main/standards.md
[commit]: https://github.com/tektoncd/community/blob/master/standards.md#commit-messages

### CODE REVIEW

**Comment by jimmyjones2** [comment]
@vinamra28 @vdemeester

File: task/buildah/0.5/buildah.yaml [hunk 1]
```diff
@@ -56,6 +56,13 @@ spec:
   - name: source
   - name: sslcertdir
     optional: true
+  - name: secrets
+    optional: true
+    description: >-
```
Comment (vinamra28) [inline task/buildah/0.5/buildah.yaml:61]: thank you for the PR @jimmyjones2, can you please document this in the readme as well?

**Comment by jimmyjones2** [comment]
Hey @vdemeester this works fine on minikube but fails on OpenShift (which is where I'd eventually like to run it - will port to tektoncd/operator once this is merged). I've seen you work at Red Hat - can you help please? If I take the ClusterTask from an OpenShift cluster to create a namespaces Task where I apply these changes I get the following error:
```
STEP 3/3: RUN --mount=type=secret,id=mysecret cat /run/secrets/mysecret
error running subprocess: error remounting "/var/tmp/buildah3094490387/mnt/rootfs/run/secrets/mysecret" in mount namespace with expected flags: permission denied
error building at STEP "RUN --mount=type=secret,id=mysecret cat /run/secrets/mysecret": exit status 1
time="2022-12-18T09:10:17Z" level=error msg="exit status 1"
```

**Comment by vdemeester** [comment]
> Hey @vdemeester this works fine on minikube but fails on OpenShift (which is where I'd eventually like to run it - will port to tektoncd/operator once this is merged). I've seen you work at Red Hat - can you help please? If I take the ClusterTask from an OpenShift cluster to create a namespaces Task where I apply these changes I get the following error:
> 
> ```
> STEP 3/3: RUN --mount=type=secret,id=mysecret cat /run/secrets/mysecret
> error running subprocess: error remounting "/var/tmp/buildah3094490387/mnt/rootfs/run/secrets/mysecret" in mount namespace with expected flags: permission denied
> error building at STEP "RUN --mount=type=secret,id=mysecret cat /run/secrets/mysecret": exit status 1
> time="2022-12-18T09:10:17Z" level=error msg="exit status 1"
> ```

Hey @jimmyjones2, yes to run this on OpenShift, you would need a "higher" set of capabilities than the default provided (through `pipeline-scc` or `restricted` scc). To be able to run this on OpenShift, you either need to grant the service account that runs this `Task` some built-in SCC like `privileged` (which is.. not really secure), or craft and attach your own SCC so that it has the required capabilities. I think it needs `CAP_SYS_ADMIN` (to be able to `mount` things — see [here](https://man7.org/linux/man-pages/man7/capabilities.7.html)).

File: task/buildah/0.5/buildah.yaml (hunk 1 above)
Comment (jimmyjones2) [reply task/buildah/0.5/buildah.yaml:61]: Done

**Comment by jimmyjones2** [comment]
> Hey @jimmyjones2, yes to run this on OpenShift, you would need a "higher" set of capabilities than the default provided (through `pipeline-scc` or `restricted` scc). To be able to run this on OpenShift, you either need to grant the service account that runs this `Task` some built-in SCC like `privileged` (which is.. not really secure), or craft and attach your own SCC so that it has the required capabilities. I think it needs `CAP_SYS_ADMIN` (to be able to `mount` things — see [here](https://man7.org/linux/man-pages/man7/capabilities.7.html)).

@vdemeester Will this PR cause issues for OpenShift users, as this Task will be [included unmodified](https://github.com/tektoncd/operator/blob/585835cfda3e618b6817bf4f1b924b96ace13d06/hack/openshift/fetch-tektoncd-catalog-tasks.sh#L47) if/when this is bumped to 0.9?

**Pull request closed without merging**